File: leetcode/q690/recursion.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        unordered_map<int, Employee*> graph;
        for (auto e : employees) {
            // graph.insert(make_pair(e->id, e));
            graph[e->id] = e;
        }
        return getImportance(id, graph);
    }
    
private:
    int getImportance(int lead, const unordered_map<int, Employee*>& graph) {
        int cur = graph.at(lead)->importance;
        for (auto sub : graph.at(lead)->subordinates) {
            cur += getImportance(sub, graph);
        }
        return cur;
    }
};

class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};

File: leetcode/q690/Recursion.java
package leetcode.q690;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

class Recursion {
    public int getImportance(List<Employee> employees, int id) {
        Map<Integer, Employee> graph = new HashMap<>();
        for (Employee e : employees) {
            graph.put(e.id, e);
        }
        
        return getImportance(id, graph);
    }
    
    private int getImportance(int root, Map<Integer, Employee> graph) {
        int cur = graph.get(root).importance;
        for (int sub : graph.get(root).subordinates) {
            cur += getImportance(sub, graph);
        }
        return cur;
    }

    static class  Employee {
        public int id;
        public int importance;
        public List<Integer> subordinates;
    }
}